File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionWeeksBetween.java
package org.hisp.dhis.rules.functions;

import org.hisp.dhis.rules.models.TimeInterval;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;
import org.joda.time.Weeks;

import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;

public class RuleFunctionWeeksBetween
    extends ScalarFunctionToEvaluate
{
    /**
     * Function which will return the number of weeks between the two given dates.
     *
     * @param start the start date.
     * @param end   the end date.
     * @return number of weeks between dates.
     */
    private Integer weeksBetween( String start, String end )
    {
        TimeInterval interval = RuleFunction.getTimeInterval( start, end );

        if ( interval.isEmpty() )
        {
            return 0;
        }

        return Weeks.weeksBetween( interval.getStartDate(), interval.getEndDate() ).getWeeks();
    }

    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        return String.valueOf(
            weeksBetween( visitor.castStringVisit( ctx.expr( 0 ) ),
                visitor.castStringVisit( ctx.expr( 1 ) ) ) );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        return null;
    }
}

File: src/jvmMain/kotlin/org/dhis2/ruleengine/functions/RuleFunctionWeeksBetween.kt
package org.dhis2.ruleengine.functions

import org.dhis2.ruleengine.functions.RuleFunction.Companion.getTimeInterval
import org.dhis2.ruleengine.parser.expression.CommonExpressionVisitor
import org.dhis2.ruleengine.parser.expression.function.ScalarFunctionToEvaluate
import org.hisp.dhis.antlr.AntlrParserUtils
import org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext
import org.joda.time.Weeks

class RuleFunctionWeeksBetween : ScalarFunctionToEvaluate() {
    /**
     * Function which will return the number of weeks between the two given dates.
     *
     * @param start the start date.
     * @param end   the end date.
     * @return number of weeks between dates.
     */
    private fun weeksBetween(start: String, end: String): Int {
        val interval = getTimeInterval(start, end)
        return if (interval.isEmpty) {
            0
        } else Weeks.weeksBetween(interval.startDate, interval.endDate).weeks
    }

    override fun evaluate(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        return weeksBetween(
            visitor.castStringVisit(ctx.expr(0)),
            visitor.castStringVisit(ctx.expr(1))
        ).toString()
    }

    override fun getDescription(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        AntlrParserUtils.castDate(visitor.visit(ctx.expr(0)))
        AntlrParserUtils.castDate(visitor.visit(ctx.expr(1)))
        return CommonExpressionVisitor.DEFAULT_DOUBLE_VALUE
    }
}